File: UICollectionViewPageLayout/UICollectionViewPageLayout.m
//
//  UICollectionViewPageLayout.m
//  UICollectionViewPageLayout
//
//  Created by Liu Bing on 9/22/15.
//  Copyright © 2015 UnixOSS. All rights reserved.
//

#import "UICollectionViewPageLayout.h"

@interface UICollectionViewPageLayout ()

@property (nonatomic, strong, nullable) NSDictionary<NSIndexPath *, UICollectionViewLayoutAttributes *> *layoutInformation;
@property (nonatomic, assign) NSInteger pageCount;

@end


@implementation UICollectionViewPageLayout

- (void)prepareLayout
{
    if (self.itemSize.width == 0 || self.itemSize.height == 0 || self.collectionView == nil) {
        return;
    }
    
    self.collectionView.pagingEnabled = YES;
    
    NSInteger rowsPerPage = self.collectionView.bounds.size.height / self.itemSize.height;
    NSInteger columnsPerPage = self.collectionView.bounds.size.width / self.itemSize.width;
    NSInteger countsPerPage = rowsPerPage * columnsPerPage;
    
    NSMutableDictionary *information = [NSMutableDictionary dictionary];
    NSInteger index = 0;
    NSInteger sectionCount = [self.collectionView numberOfSections];
    for (NSInteger section = 0; section < sectionCount; section++) {
        NSInteger itemCount = [self.collectionView numberOfItemsInSection:section];
        for (NSInteger item = 0; item < itemCount; item++) {
            NSIndexPath *indexPath = [NSIndexPath indexPathForItem:item inSection:section];
            NSInteger page = index / countsPerPage;
            NSInteger count = index % countsPerPage;
            NSInteger row = count / columnsPerPage;
            NSInteger column = count % columnsPerPage;
            information[indexPath] = [self attributesAtIndexPath:indexPath page:page row:row column:column];
            index++;
        }
    }
    self.layoutInformation = [information copy];
    self.pageCount = (index % countsPerPage != 0) ? index/countsPerPage+1 : index/countsPerPage;
}

- (UICollectionViewLayoutAttributes *)attributesAtIndexPath:(NSIndexPath *)indexPath page:(NSInteger)page row:(NSInteger)row column:(NSInteger)column
{
    CGFloat originX = page * self.collectionView.bounds.size.width + column * self.itemSize.width;
    CGFloat originY = self.itemSize.height * row;
    UICollectionViewLayoutAttributes *attributes = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
    attributes.frame = CGRectMake(originX, originY, self.itemSize.width, self.itemSize.height);
    return attributes;
}

- (CGSize)collectionViewContentSize
{
    return CGSizeMake(self.collectionView.bounds.size.width * self.pageCount, self.collectionView.bounds.size.height);
}

- (NSArray<__kindof UICollectionViewLayoutAttributes *> * _Nullable)layoutAttributesForElementsInRect:(CGRect)rect
{
    __block NSMutableArray *attributes = [NSMutableArray arrayWithCapacity:self.layoutInformation.count];
    [self.layoutInformation enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, UICollectionViewLayoutAttributes * _Nonnull obj, BOOL * _Nonnull stop) {
        if (CGRectIntersectsRect(rect, obj.frame)) {
            [attributes addObject:obj];
        }
    }];
    return [attributes copy];
}

- (UICollectionViewLayoutAttributes * _Nullable)layoutAttributesForItemAtIndexPath:(NSIndexPath * _Nonnull)indexPath
{
    return self.layoutInformation[indexPath];
}

@end

File: UICollectionViewPageLayout/UICollectionViewPageLayout.swift
//
//  UICollectionViewPageLayout.swift
//  UICollectionViewPageLayout
//
//  Created by Liu Bing on 9/25/15.
//  Copyright © 2015 UnixOSS. All rights reserved.
//

import UIKit

public class UICollectionViewPageLayout: UICollectionViewLayout {
    
    public var itemSize: CGSize = CGSizeZero
    
    private var layoutInformation = [NSIndexPath: UICollectionViewLayoutAttributes]()
    private var pageCount = 0
    
    public override func prepareLayout() {
        if itemSize.width == 0 || itemSize.height == 0 || collectionView == nil {
            return
        }
        let rowsPerPage = Int(collectionView!.bounds.size.height / itemSize.height)
        let columnsPerPage = Int(collectionView!.bounds.size.width / itemSize.width)
        let countsPerPage = rowsPerPage * columnsPerPage
        
        collectionView?.pagingEnabled = true
        var _layoutInformation = [NSIndexPath: UICollectionViewLayoutAttributes]()
        var index = 0
        let sectionCount = collectionView?.numberOfSections() ?? 0
        for section in 0..<sectionCount {
            let itemCount = collectionView?.numberOfItemsInSection(section) ?? 0
            for item in 0..<itemCount {
                let indexPath = NSIndexPath(forItem: item, inSection: section)
                let page = index / countsPerPage
                let count = index % countsPerPage
                let row = count / columnsPerPage
                let column = count % columnsPerPage
                _layoutInformation[indexPath] = attributesAtIndexPath(indexPath, page: page, row: row, column: column)
                index++
            }
        }
        layoutInformation = _layoutInformation
        pageCount = (index % countsPerPage != 0) ? index / countsPerPage + 1 : index / countsPerPage
    }
    
    private func attributesAtIndexPath(indexPath: NSIndexPath, page: Int, row: Int, column: Int) -> UICollectionViewLayoutAttributes {
        let originX = CGFloat(page) * (collectionView?.bounds.size.width ?? 0) + CGFloat(column) * itemSize.width
        let originY = itemSize.height * CGFloat(row)
        let attributes = UICollectionViewLayoutAttributes(forCellWithIndexPath: indexPath)
        attributes.frame = CGRect(x: originX, y: originY, width: itemSize.width, height: itemSize.height)
        return attributes
    }
    
    public override func collectionViewContentSize() -> CGSize {
        let collectionViewSize = collectionView?.bounds.size ?? CGSizeZero
        return CGSize(width: collectionViewSize.width * CGFloat(pageCount), height: collectionViewSize.height)
    }
    
    public override func layoutAttributesForElementsInRect(rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        var attributes = [UICollectionViewLayoutAttributes]()
        for (_, value) in layoutInformation {
            if CGRectIntersectsRect(rect, value.frame) {
                attributes.append(value)
            }
        }
        return attributes
    }
    
    public override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
        return layoutInformation[indexPath]
    }
    
}
